File: Linked List/NthNodeFromEndOfLinkedList.py
__author__= 'Kalpak Seal'

'''
Process:
Maintain two pointers – reference pointer and main pointer. Initialize both reference and main pointers to head. 
First move reference pointer to n nodes from head. Now move both pointers one by one until reference pointer reaches end. 
Now main pointer will point to nth node from the end. 
Return main pointer.
'''
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

    def get_data(self):
        return self.data


class LinkedList:
    def __init__(self):
        self.head = None

    def insert_at_beginning(self, data):
        new_node = Node(data)
        new_node.next = self.head
        self.head = new_node

    def insert_after_a_node(self, prev_node, data):
        if prev_node is None:
            print("Not a Valid previous node")
            return

        new_node = Node(data)
        new_node.next = prev_node.next
        prev_node.next = new_node

    def insert_at_end(self, data):
        new_node = Node(data)

        if self.head is None:
            self.head = new_node
            return

        last_node = self.head
        while last_node.next is not None:
            last_node = last_node.next

        last_node.next = new_node

def nth_node_from_end_of_linked_list(head, n):
    main_ptr = head
    ref_ptr = head

    count = 0

    if head is not None:
        while count < n:
            if ref_ptr is None:
                print(n, " is greater than the number of nodes in the linked list")
                return
            ref_ptr = ref_ptr.next
            count += 1

    while ref_ptr is not None:
        ref_ptr = ref_ptr.next
        main_ptr = main_ptr.next

    print ("The %dth node from the end of the linked list is : %d" %(n, main_ptr.get_data()))


def print_linked_list(head):
    temp = head
    while temp:
        print(temp.get_data(), end=' ')
        temp = temp.next
    print()


if __name__ == '__main__':
    llist = LinkedList()
    llist.head = Node(1)
    second = Node(2)
    third = Node(3)

    llist.head.next = second
    second.next = third

    llist.insert_at_beginning(4)
    llist.insert_at_end(5)
    llist.insert_after_a_node(llist.head.next, 6)

    print("The linked list is: ")
    print_linked_list(llist.head)

    nth_node_from_end_of_linked_list(llist.head, 3)

File: Linked List/NthNodeFromEndOfLinkedList.c
#include <stdio.h>
#include <stdlib.h>

struct Node {
	int data;
	struct Node* next;
};

void insert_at_end(struct Node** head_ref, int data){
	struct Node* current_node = *head_ref;
 
	struct Node* new_node = (struct Node*)malloc(sizeof(struct Node));
	new_node->data = data;
	new_node->next = NULL;
 
	if (*head_ref == NULL){
		*head_ref = new_node;
		return;
	}
 
	while(current_node->next != NULL){
		current_node = current_node->next;
	}
 
	current_node->next = new_node;
	return;
 
}

void printLinkedList(struct Node* head){
	while (head != NULL){
		printf("%d ", head -> data);
		head = head -> next;
	}
	printf("\n");
}

void printNthNodeFromEnd(struct Node* head_ref, int n){
	struct Node* main_ptr = head_ref;
	struct Node* ref_ptr = head_ref;
	
	int count = 0;
	
	if (head_ref != NULL){
		while (count < n){
			if (ref_ptr == NULL){
				printf("%d greater than no. of nodes in the linked list", n);
				return;
			}
			ref_ptr = ref_ptr->next;
			count ++;
		}
		
		while (ref_ptr != NULL){
			main_ptr = main_ptr->next;
			ref_ptr = ref_ptr-> next;
		}
		printf("Node no. %d from last is %d ",n, main_ptr->data);
	}
}

int main(void) {
	struct Node* head = NULL;
    insert_at_end(&head, 1);
    insert_at_end(&head, 2);
    insert_at_end(&head, 3);
    insert_at_end(&head, 4);
    insert_at_end(&head, 5);
    
    printLinkedList(head);
    
    printNthNodeFromEnd(head, 3);
    
	return 0;
}
